clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -109,78 +112,72 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+                            throw new TimeoutException("Failed to allocate memory within the configured max blocking time " + maxTimeToBlockMs + " ms.");
+                        }
 
-                // remove the condition for this thread to let the next thread
-                // in line start getting memory
-                Condition removed = this.waiters.removeFirst();
-                if (removed != moreMemory)
-                    throw new IllegalStateException("Wrong condition: this shouldn't happen.");
+                        remainingTimeToBlockNs -= timeNs;
 
-                // signal any additional waiters if there is more memory left
-                // over for them
-                if (this.availableMemory > 0 || !this.free.isEmpty()) {
-                    if (!this.waiters.isEmpty())
-                        this.waiters.peekFirst().signal();
-                }
+                        // check if we can satisfy this request from the free list,
+                        // otherwise allocate memory
+                        if (accumulated == 0 && size == this.poolableSize && !this.free.isEmpty()) {
+                            // just grab a buffer from the free list
+                            buffer = this.free.pollFirst();
+                            accumulated = size;
+                        } else {
+                            // we'll need to allocate memory, but we may only get
+                            // part of what we need on this iteration
+                            freeUp(size - accumulated);
+                            int got = (int) Math.min(size - accumulated, this.availableMemory);
+                            this.availableMemory -= got;
+                            accumulated += got;
+                        }
+                    }
 
-                // unlock and return the buffer
-                lock.unlock();
-                if (buffer == null)
-                    return allocateByteBuffer(size);
-                else
+                    if (buffer == null)
+                        buffer = allocateByteBuffer(size);
+                    hasError = false;
+                    //unlock happens in top-level, enclosing finally
                     return buffer;
+                } finally {
+                    // When this loop was not able to successfully terminate don't loose available memory
+                    if (hasError)
+                        this.availableMemory += accumulated;
+                    this.waiters.remove(moreMemory);
+                }
             }
         } finally {
-            if (lock.isHeldByCurrentThread())
-                lock.unlock();
+            // signal any additional waiters if there is more memory left
+            // over for them
+            if (!(this.availableMemory == 0 && this.free.isEmpty()) && !this.waiters.isEmpty())
+                this.waiters.peekFirst().signal();
+            lock.unlock();
```

Comment:
It seems findBugs is not happy with this finally block although I think it should be fine. Can we have a try finally block here to ensure lock.unlock() is executed so that findBugs is happy?